src/txn/txn_ckpt.c

```diff
@@ -1439,12 +1440,29 @@ __checkpoint_lock_dirty_tree(
         }
     }
 
+    /*
+     * Discard the saved list of checkpoints, and slow path if this is not a WiredTiger checkpoint
+     * or if checkpoint drops are involved. Also, if we do not have checkpoint array size, the
+     * regular checkpoint process did not create the array. It is safer to discard the array in such
+     * a case.
+     */
+    if (!is_wt_ckpt || is_drop || btree->ckpt_allocated == 0)
+        __wt_meta_saved_ckptlist_free(session);
+
```

Comment:
@agorrod and I were talking recently that maybe it's time to simply drop support for anything other than database-wide checkpoints. Now that you've wrapped your head around the checkpoint code, maybe a good follow-on ticket would be to remove that code.

Reply:
That would have made this change a bit simpler, and easy to get around my head the first time. I have created WT-7509 for further discussion. Please feel free to fill in more details. Do you also mean to drop support for named checkpoints in such a case?

Reply:
I don't think we want to drop support for named checkpoints, there's still a use for them, I think -- only dropping partial checkpoints.